Timing.repeat.boundsPeriod without end time causes NullPointerException in BaseSearchParamExtractor.addDate_Timing

**Describe the bug**
NullPointerExcection when updating a resource with Timing.repeat.boundsPeriod without end time 

**To Reproduce**
```
        Period period = new Period();
        period.setStart(new Date());
        Timing timing = new Timing();
        timing.setRepeat(new Timing.TimingRepeatComponent().setBounds(period));
        procedureRequest.setOccurrence(timing);
        genericClient.update().resource(procedureRequest).execute();

```
Observed exception:
```
 Failed to call access method: java.lang.NullPointerException
java.lang.NullPointerException: null
        at java.util.TreeMap.put(TreeMap.java:561) ~[?:?]
        at java.util.TreeSet.add(TreeSet.java:255) ~[?:?]
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.addDate_Timing(BaseSearchParamExtractor.java:635) ~[hapi-fhir-jpaserver-searchparam-5.0.2.jar:?]
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.lambda$extractSearchParamDates$3(BaseSearchParamExtractor.java:320) ~[hapi-fhir-jpaserver-searchparam-5.0.2.jar:?]
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.extractSearchParam(BaseSearchParamExtractor.java:814) ~[hapi-fhir-jpaserver-searchparam-5.0.2.jar:?]
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.extractSearchParams(BaseSearchParamExtractor.java:796) ~[hapi-fhir-jpaserver-searchparam-5.0.2.jar:?]
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.extractSearchParamDates(BaseSearchParamExtractor.java:331) ~[hapi-fhir-jpaserver-searchparam-5.0.2.jar:?]
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.extractSearchParamDates(SearchParamExtractorService.java:367) ~[hapi-fhir-jpaserver-searchparam-5.0.2.jar:?]
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.extractSearchIndexParameters(SearchParamExtractorService.java:120) ~[hapi-fhir-jpaserver-searchparam-5.0.2.jar:?]
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.extractFromResource(SearchParamExtractorService.java:94) ~[hapi-fhir-jpaserver-searchparam-5.0.2.jar:?]

```

**Environment (please complete the following information):**
 - HAPI FHIR Version: 5.0.2

Hello, I face somehow a similar issue with HAPI FHIR version 4.1.0.
Posting a resource ServiceRequest like below with a boundsPeriod

> {
>     "resourceType": "ServiceRequest",
>     "status": "active",
>     "code": {
>         "coding": [
>             {
>                 "system": "http://loinc.org",
>                 "code": "2880-3"
>             }
>         ],
>         "text": "Protéines LCR"
>     },
>     "quantityQuantity": {"value": 1},
>     "subject": {"reference": "Patient/18454"},
>     "occurrenceTiming": {
>         "repeat": {
>             "boundsPeriod": {"start": "2021-01-24"},
>             "count": 0,
>             "duration": 15,
>             "durationUnit": "d",
>             "period": 3,
>             "periodUnit": "d"
>         }
>     },
>     "requester": {"reference": "Practitioner/18448"}
> }

 makes a NullPointerException in the HAPI server as shown below

`Caused by: java.lang.reflect.InvocationTargetException: null
        at sun.reflect.GeneratedMethodAccessor223.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at ca.uhn.fhir.rest.server.method.BaseMethodBinding.invokeServerMethod(BaseMethodBinding.java:239)
        ... 40 common frames omitted
Caused by: java.lang.NullPointerException: null
        at java.util.TreeMap.put(TreeMap.java:563)
        at java.util.TreeSet.add(TreeSet.java:255)
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.addDate_Timing(BaseSearchParamExtractor.java:582)
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.lambda$extractSearchParamDates$2(BaseSearchParamExtractor.java:263)
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.extractSearchParams(BaseSearchParamExtractor.java:722)
        at ca.uhn.fhir.jpa.searchparam.extractor.BaseSearchParamExtractor.extractSearchParamDates(BaseSearchParamExtractor.java:274)
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.extractSearchParamDates(SearchParamExtractorService.java:121)
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.extractFromResource(SearchParamExtractorService.java:62)
        at ca.uhn.fhir.jpa.dao.index.SearchParamWithInlineReferencesExtractor.populateFromResource(SearchParamWithInlineReferencesExtractor.java:92)
        at ca.uhn.fhir.jpa.dao.BaseHapiFhirDao.updateEntity(BaseHapiFhirDao.java:982)
        at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.doCreate(BaseHapiFhirResourceDao.java:451)
        at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.create(BaseHapiFhirResourceDao.java:163)
        at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.create(BaseHapiFhirResourceDao.java:134)
        at sun.reflect.GeneratedMethodAccessor224.invoke(Unknown Source)
`

--- a/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_4_0/2417-avoid-npe-indexing-timing.yaml
+++ b/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_4_0/2417-avoid-npe-indexing-timing.yaml
@@ -1,0 +1,5 @@
+---
+type: fix
+issue: 2417
+title: "A NullPointerException was corrected when indexing resources containing an indexed Period field that
+  had a start but not an end defined."

--- a/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java
+++ b/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java
@@ -1497,6 +1497,43 @@ public class FhirResourceDaoR4SearchNoFtTest extends BaseJpaR4Test {
 
 
 	@Test
+	public void testPeriodWithNoStart() {
+		ServiceRequest serviceRequest = new ServiceRequest();
+
+		Period period = new Period();
+		period.setEnd(new Date());
+		Timing timing = new Timing();
+		timing.setRepeat(new Timing.TimingRepeatComponent().setBounds(period));
+		serviceRequest.setOccurrence(timing);
+
+		// Should not crash
+		myServiceRequestDao.create(serviceRequest);
+
+		runInTransaction(()->{
+			assertEquals(1, myResourceIndexedSearchParamDateDao.findAll().size());
+		});
+	}
+
+	@Test
+	public void testPeriodWithNoEnd() {
+		ServiceRequest serviceRequest = new ServiceRequest();
+		
+		Period period = new Period();
+		period.setStart(new Date());
+		Timing timing = new Timing();
+		timing.setRepeat(new Timing.TimingRepeatComponent().setBounds(period));
+		serviceRequest.setOccurrence(timing);
+
+		// Should not crash
+		myServiceRequestDao.create(serviceRequest);
+
+		runInTransaction(()->{
+			assertEquals(1, myResourceIndexedSearchParamDateDao.findAll().size());
+		});
+	}
+	
+
+	@Test
 	public void testSearchByIdParamInverse() {
 		String id1;
 		{

--- a/hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java
+++ b/hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java
@@ -1364,8 +1364,12 @@ public abstract class BaseSearchParamExtractor implements ISearchParamExtractor 
 					if ("Period".equals(boundsType)) {
 						Date start = extractValueAsDate(myPeriodStartValueChild, bounds.get());
 						Date end = extractValueAsDate(myPeriodEndValueChild, bounds.get());
-						dates.add(start);
+						if (start != null) {
+							dates.add(start);
+						}
+						if (end != null) {
 						dates.add(end);
+						}
 					}
 				}
 			}

